File: examples/order/python/src/orders/service.py
import uuid
from typing import Dict, List, Optional
from .schemas import Order, OrderCreate, OrderUpdate, OrderStatus
from ..common.errors import NotFoundError, ConflictError
from ..events.event_service import event_service
import structlog

logger = structlog.get_logger(__name__)


class OrderService:
    """Service for managing orders."""
    
    def __init__(self):
        # In-memory storage for demo purposes
        # In production, this would be a database
        self._orders: Dict[str, Order] = {}
        logger.info("OrderService initialized")
    
    async def create_order(self, order_data: OrderCreate) -> Order:
        """Create a new order."""
        order_id = str(uuid.uuid4())
        
        order = Order(
            id=order_id,
            status=OrderStatus.PENDING,
            customerId=order_data.customerId,
            items=order_data.items
        )
        
        self._orders[order_id] = order
        
        # Publish order created event
        event_service.publish_order_created(
            order_id=order.id,
            customer_id=order.customerId,
            items=order.items,
            status=order.status.value
        )
        
        logger.info(
            "Order created",
            order_id=order.id,
            customer_id=order.customerId,
            items_count=len(order.items),
            status=order.status
        )
        
        return order
    
    async def get_order(self, order_id: str) -> Order:
        """Get an order by ID."""
        order = self._orders.get(order_id)
        if not order:
            raise NotFoundError(f"Order with ID {order_id} not found")
        
        logger.debug("Order retrieved", order_id=order_id, status=order.status)
        return order
    
    async def update_order(self, order_id: str, update_data: OrderUpdate) -> Order:
        """Update an order's status."""
        order = self._orders.get(order_id)
        if not order:
            raise NotFoundError(f"Order with ID {order_id} not found")
        
        # Validate status transition
        self._validate_status_transition(order.status, update_data.status)
        
        previous_status = order.status
        order.status = update_data.status
        
        # Publish order updated event
        event_service.publish_order_updated(
            order_id=order.id,
            status=order.status.value,
            previous_status=previous_status.value
        )
        
        logger.info(
            "Order status updated",
            order_id=order.id,
            previous_status=previous_status,
            new_status=order.status
        )
        
        return order
    
    async def list_orders(self, limit: int = 10, offset: int = 0) -> tuple[List[Order], int]:
        """List orders with pagination."""
        all_orders = list(self._orders.values())
        total = len(all_orders)
        
        # Sort by creation order (in real app, you'd sort by creation timestamp)
        orders_page = all_orders[offset:offset + limit]
        
        logger.debug(
            "Orders listed",
            total=total,
            limit=limit,
            offset=offset,
            returned=len(orders_page)
        )
        
        return orders_page, total
    
    def _validate_status_transition(self, current_status: OrderStatus, new_status: OrderStatus) -> None:
        """Validate that the status transition is allowed."""
        # Define allowed transitions
        transitions = {
            OrderStatus.PENDING: [OrderStatus.PAID],
            OrderStatus.PAID: [OrderStatus.SHIPPED],
            OrderStatus.SHIPPED: [OrderStatus.DELIVERED],
            OrderStatus.DELIVERED: []  # Terminal state
        }
        
        allowed_statuses = transitions.get(current_status, [])
        
        if new_status not in allowed_statuses:
            raise ConflictError(
                f"Cannot update order status from {current_status.value} to {new_status.value}"
            )


# Global service instance
order_service = OrderService()

File: examples/order/rust/src/orders/service.rs
use chrono::Utc;
use std::{collections::HashMap, sync::Arc};
use tokio::sync::RwLock;
use tracing::{info, instrument};
use uuid::Uuid;

use crate::{
    common::errors::AppError,
    events::EventService,
};

use super::{CreateOrderDto, Order, OrderStatus, UpdateOrderDto};

#[derive(Clone)]
pub struct OrdersService {
    // In-memory storage for demo purposes
    // In production, this would be a database
    orders: Arc<RwLock<HashMap<String, Order>>>,
    event_service: Arc<EventService>,
}

impl OrdersService {
    pub fn new(event_service: Arc<EventService>) -> Self {
        Self {
            orders: Arc::new(RwLock::new(HashMap::new())),
            event_service,
        }
    }

    #[instrument(skip(self))]
    pub async fn create_order(&self, dto: CreateOrderDto) -> Result<Order, AppError> {
        let now = Utc::now();
        let order = Order {
            id: Uuid::new_v4().to_string(),
            customer_id: dto.customer_id,
            items: dto.items,
            status: OrderStatus::Pending,
            created_at: now,
            updated_at: now,
        };

        // Store the order
        {
            let mut orders = self.orders.write().await;
            orders.insert(order.id.clone(), order.clone());
        }

        // Publish event
        if let Err(e) = self.event_service.publish_order_created(&order).await {
            info!(error = %e, order_id = %order.id, "Failed to publish order created event");
        }

        info!(order_id = %order.id, customer_id = %order.customer_id, "Order created");
        Ok(order)
    }

    #[instrument(skip(self))]
    pub async fn get_order(&self, id: &str) -> Result<Order, AppError> {
        let orders = self.orders.read().await;
        orders
            .get(id)
            .cloned()
            .ok_or(AppError::OrderNotFound)
    }

    #[instrument(skip(self))]
    pub async fn update_order(&self, id: &str, dto: UpdateOrderDto) -> Result<Order, AppError> {
        let mut orders = self.orders.write().await;
        
        let order = orders
            .get_mut(id)
            .ok_or(AppError::OrderNotFound)?;

        let old_status = order.status.clone();
        
        // Update fields if provided
        if let Some(status) = dto.status {
            // Validate status transitions
            match (&order.status, &status) {
                (OrderStatus::Pending, OrderStatus::Paid) => {},
                (OrderStatus::Paid, OrderStatus::Shipped) => {},
                (OrderStatus::Shipped, OrderStatus::Delivered) => {},
                _ => {
                    return Err(AppError::OrderConflict(
                        format!("Invalid status transition from {} to {}", old_status, status)
                    ));
                }
            }
            
            order.status = status;
        }

        order.updated_at = Utc::now();
        let updated_order = order.clone();

        // Release the lock before publishing the event
        drop(orders);

        // Publish event if status changed
        if old_status.to_string() != updated_order.status.to_string() {
            if let Err(e) = self.event_service.publish_order_updated(&updated_order).await {
                info!(error = %e, order_id = %updated_order.id, "Failed to publish order updated event");
            }
        }

        info!(
            order_id = %updated_order.id,
            old_status = %old_status,
            new_status = %updated_order.status,
            "Order updated"
        );

        Ok(updated_order)
    }

    #[instrument(skip(self))]
    pub async fn list_orders(&self, limit: Option<u32>, offset: Option<u32>) -> Result<Vec<Order>, AppError> {
        let orders = self.orders.read().await;
        let mut order_list: Vec<Order> = orders.values().cloned().collect();
        
        // Sort by created_at descending (newest first)
        order_list.sort_by(|a, b| b.created_at.cmp(&a.created_at));

        // Apply pagination
        let offset = offset.unwrap_or(0) as usize;
        let limit = limit.unwrap_or(10) as usize;

        let result = order_list
            .into_iter()
            .skip(offset)
            .take(limit)
            .collect();

        Ok(result)
    }
}
